lib/SIL/IR/SILInstruction.cpp

```diff
@@ -1295,17 +1295,57 @@ bool SILInstruction::isDeallocatingStack() const {
 bool SILInstruction::mayRequirePackMetadata() const {
   switch (getKind()) {
   case SILInstructionKind::AllocPackInst:
-  case SILInstructionKind::PartialApplyInst:
-  case SILInstructionKind::ApplyInst:
-  case SILInstructionKind::BeginApplyInst:
-  case SILInstructionKind::TryApplyInst:
-  case SILInstructionKind::DebugValueInst:
-  case SILInstructionKind::MetatypeInst:
   case SILInstructionKind::TuplePackElementAddrInst:
   case SILInstructionKind::OpenPackElementInst:
-  case SILInstructionKind::ClassMethodInst:
-  case SILInstructionKind::WitnessMethodInst:
     return true;
+  case SILInstructionKind::PartialApplyInst:
+  case SILInstructionKind::ApplyInst:
+  case SILInstructionKind::BeginApplyInst:
+  case SILInstructionKind::TryApplyInst: {
+    // Check the function type for packs.
+    auto apply = ApplySite::isa(const_cast<SILInstruction *>(this));
+    if (apply.getCallee()->getType().isOrContainsPack(*getFunction()))
+      return true;
+    // Check the function signature for packs.
+    auto signature = apply.getOrigCalleeType()->getInvocationGenericSignature();
+    if (!signature)
+      return false;
+    auto result = false;
+    signature->forEachParam(([&](GenericTypeParamType *ty, bool canonical) {
+      // We already found an interesting parameter, don't bother looking at the
+      // rest.
+      if (result)
+        return;
+      if (!canonical)
+        return;
+      auto *function = getFunction();
+      auto loweredTy =
```

Comment:
Lowering isn't ever going to change a generic parameter type. All of this can be simplified to `if (ty->isParameterPack()) result = true;